[FEATURE]: Adding an XML Catalog file

Nicolas-Peiffer opened the pull request:
## Describe the feature

Adding an XML catalog `schema/xmlcatalog.xml` file under the `schema/` folder. This addition is transparent, but will allow other projects to use this XML catalog in their XML parser.

```xml
<?xml version="1.0"?>
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
  <uri name="http://cyclonedx.org/schema/spdx" uri="spdx.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.0" uri="bom-1.0.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.1" uri="bom-1.1.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.2" uri="bom-1.2.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.3" uri="bom-1.3.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.4" uri="bom-1.4.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.5" uri="bom-1.5.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.6" uri="bom-1.6.xsd"/>
</catalog>
```

## Possible solutions

While working with `cyclonedx-go`, I noticed 2 variables `xmlNamespaces` and `xmlSchemaFiles`:

* [xmlNamespaces](https://github.com/CycloneDX/cyclonedx-go/blob/98a070df0171240c53e7e1c3c46640eb9b257e08/cyclonedx_xml.go#L409-L417): which is used to bind XML namespaces URLs with CycloneDX spec versions; 
* [xmlSchemaFiles](https://github.com/CycloneDX/cyclonedx-go/blob/98a070df0171240c53e7e1c3c46640eb9b257e08/validate_xml_test.go#L27C1-L36C1): which is used to bind local XSD schema files location path with CycloneDX spec versions; 

```go
var xmlSchemaFiles = map[SpecVersion]string{
	SpecVersion1_0: "./schema/bom-1.0.xsd",
	SpecVersion1_1: "./schema/bom-1.1.xsd",
	SpecVersion1_2: "./schema/bom-1.2.xsd",
	SpecVersion1_3: "./schema/bom-1.3.xsd",
	SpecVersion1_4: "./schema/bom-1.4.xsd",
	SpecVersion1_5: "./schema/bom-1.5.xsd",
	SpecVersion1_6: "./schema/bom-1.6.xsd",
}
```

```go
var xmlNamespaces = map[SpecVersion]string{
	SpecVersion1_0: "http://cyclonedx.org/schema/bom/1.0",
	SpecVersion1_1: "http://cyclonedx.org/schema/bom/1.1",
	SpecVersion1_2: "http://cyclonedx.org/schema/bom/1.2",
	SpecVersion1_3: "http://cyclonedx.org/schema/bom/1.3",
	SpecVersion1_4: "http://cyclonedx.org/schema/bom/1.4",
	SpecVersion1_5: "http://cyclonedx.org/schema/bom/1.5",
	SpecVersion1_6: "http://cyclonedx.org/schema/bom/1.6",
}
```

They were lacking information about `<uri name="http://cyclonedx.org/schema/spdx" uri="spdx.xsd"/>`.

My goal is to write an XML validation method for XML formated BOMs. I am discussing this possibility on the [CycloneDX Go Slack channel](https://cyclonedx.slack.com/archives/C01PZRT73K9/p1717767294262849?thread_ts=1717695511.115299&cid=C01PZRT73K9). Especially, we are discussing an issue about the need to enable `CGO` when using `github.com/lestrrat-go/libxml2`.

XML Catalogs are very handy when it comes to solving the issue of binding XML Namespace URLs with locally stored XSD files.

See my example here:
https://github.com/Nicolas-Peiffer/cyclonedx-go/blob/427e8adad7e05f57008c78a76075ddc6e38fb007/validator_xml.go#L47-L48

```go
import (
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

[...]

	// Load the XSD schema and the XML catalog
	schema, err := xsd.Parse(schemaData.Bytes(), xsd.WithPath("schema/xmlcatalog.xml"))
```

First tests of using an XML catalog seems promising: the tests go through with success.
https://github.com/Nicolas-Peiffer/cyclonedx-go/commit/a4f347065cba19e4e41ec16cc652913e23e221f5

![Copie d'écran_20240607_190900](https://github.com/CycloneDX/specification/assets/102670102/7a720498-94a2-4bf5-b727-662124381f6c)


## Alternatives

I do not know alternatives to XML catalogs, except implementing custom variables like `cyclonedx-go`'s `xmlNamespaces` and `xmlSchemaFiles` variables.

## Additional context

See the Slack Go channel.

jkowalleck commented:
maybe add some regression test that checks whether the namespace in the catalog matches the referenced file, and assert that the referenced file exists.

Nicolas-Peiffer commented on schema/xmlcatalog.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
```
Short answer: we do not need DTD when we already use XSD schema files.
But I am not an XML expert, I put below the result of a question to chatGPT about whether a DTD should be included in this particular XML catalog for CycloneDX or not.

Is this answer is satisfactory?

***

A Document Type Definition (DTD) is used to define the structure and the legal elements and attributes of an XML document. While DTDs can be useful for ensuring the correctness and validity of an XML document, they are not strictly necessary for all XML documents, especially if you are using XML Schema Definitions (XSD) instead.

In the context of an XML catalog, which is primarily used to map public identifiers and system identifiers to local resources, a DTD is not typically required. XML catalogs are more about resolving URIs to local files rather than enforcing a document structure.

However, if you want to ensure that your XML catalog adheres to a specific structure and follows the XML Catalog standard, you can refer to the DTD for XML catalogs. The XML Catalog specification itself defines a DTD that can be used.

Here’s an example of how you can reference the XML Catalog DTD in your XML catalog file:

```xml
<?xml version="1.0"?>
<!DOCTYPE catalog PUBLIC "-//OASIS//DTD XML Catalogs V1.1//EN" "http://www.oasis-open.org/committees/entity/release/1.1/catalog.dtd">
<!--
  This XML catalog provides mappings for CycloneDX schemas.
  The catalog maps schema URLs to local XSD files to facilitate schema validation without needing internet access.
  Namespace: urn:oasis:names:tc:entity:xmlns:xml:catalog
-->
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">

  <!-- SPDX Schema -->
  <uri name="http://cyclonedx.org/schema/spdx" uri="spdx.xsd"/>

  <!-- BOM Schemas -->
  <uri name="http://cyclonedx.org/schema/bom/1.0" uri="bom-1.0.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.1" uri="bom-1.1.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.2" uri="bom-1.2.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.3" uri="bom-1.3.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.4" uri="bom-1.4.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.5" uri="bom-1.5.xsd"/>
  <uri name="http://cyclonedx.org/schema/bom/1.6" uri="bom-1.6.xsd"/>

  <!-- Placeholder for future schemas -->
  <!-- <uri name="http://cyclonedx.org/schema/bom/1.x" uri="bom-1.x.xsd"/> -->

</catalog>
```

### Explanation

1. **DOCTYPE Declaration**:
   - The `<!DOCTYPE catalog PUBLIC "-//OASIS//DTD XML Catalogs V1.1//EN" "http://www.oasis-open.org/committees/entity/release/1.1/catalog.dtd">` declaration specifies that this document adheres to the XML Catalog DTD.

2. **Comments**:
   - Comments explain the purpose of the XML catalog and the schema mappings.

### When to Use DTD

Use a DTD if:
- You need to enforce a specific structure in your XML documents.
- You want to validate the XML against the DTD to catch structural errors.

For XML catalogs specifically, adhering to the XML Catalog DTD can help ensure that your catalog files are correct and compliant with the standard.

### When Not to Use DTD

If you are already using XSDs to validate your XML documents and your toolchain supports XSD validation well, you might not need a DTD. XSDs offer more powerful validation features compared to DTDs.

### Summary

While you don't strictly need a DTD for an XML catalog, referencing the XML Catalog DTD can help ensure your catalog file is well-formed and adheres to the standard. This can be particularly useful if the catalog will be used in environments that validate against the DTD. If your use case benefits from this validation, including the DTD reference is a good practice.

jkowalleck replied:
see https://github.com/ThalesGroup/cyclonedx-specification/pull/1

Nicolas-Peiffer replied:
I merged and added your comment in the XML catalog, thank you. I added a couple of other comments as well.

https://github.com/ThalesGroup/cyclonedx-specification/commit/4631813151b94f76dd8aa6b536216b48e779b048

Nicolas-Peiffer commented:
> maybe add some sanity/plausibility/regression test that checks whether the namespace in the catalog matches the referenced file, and assert that the referenced file exists.

Good idea. But first, I updated the version of `cyclonedx-core-java` from `cyclonedx-core-java-8.0.3` to `cyclonedx-core-java-9.0.2`, because `9.0.2` is the latest available: see my new [`pom.xml` file](https://github.com/ThalesGroup/cyclonedx-specification/blob/3400f757c02753686863695d06282173951da8e6/tools/pom.xml#L58).

And I added the tests for CDX v1.6: indeed you can see below that 1.6 was not tested:
* [JsonSchemaVerificationTest.java](https://github.com/CycloneDX/specification/blob/491e8138c3a37e1da406d49102320e8de32897e0/tools/src/test/java/org/cyclonedx/schema/JsonSchemaVerificationTest.java#L43) only goes to 1.5
* [BaseSchemaVerificationTest.java](https://github.com/CycloneDX/specification/blob/491e8138c3a37e1da406d49102320e8de32897e0/tools/src/test/java/org/cyclonedx/schema/BaseSchemaVerificationTest.java#L31) only goes to 1.5
* [XmlSchemaVerificationTest.java](https://github.com/CycloneDX/specification/blob/491e8138c3a37e1da406d49102320e8de32897e0/tools/src/test/java/org/cyclonedx/schema/XmlSchemaVerificationTest.java#L47) only goes to 1.5

So far, the tests seems to go through for 1.6, which is a good news :smiley: .

Now I need to implement tests for the `xmlcatalog.xml` file like you suggested. This is a work in progress. I have some trouble locating the file under `./schema/` when the test source code is here `./tools/src/test/java/org/cyclonedx/schema/`. I need to figure this out.

NB : if I were doing my tool in Java, it seems the CycloneDX Java implementation is more mature than the `cyclonedx-go`.

![image](https://github.com/CycloneDX/specification/assets/102670102/2b37f4ac-526c-4663-8abb-e3f00bafd4b4)

Nicolas-Peiffer reviewed (commented):
I change the pr status to draft again, because we need to solve the XMLCatalog tests firts.

Nicolas-Peiffer commented:
@jkowalleck My latest commit https://github.com/ThalesGroup/cyclonedx-specification/commit/c1fe8a9781ee08d766292bdf269391e909712707 seems very promising:

* I added your comments on the XML catalog, plus a couple of other comments;
* I managed to write a Java test for testing the content of the XML Catalog and verify if the content of the XSD files match
* I updated the JSON and XML Java test to CycloneDX 1.6
* Java unit tests go trough and validate !

And on my machine environment (VSCode, openjdk 22), tests seems to run correctly. But I have no way of testing on another environment. Unless I activate github actions...

```
java  --version

openjdk 22 2024-03-19
OpenJDK Runtime Environment (build 22)
OpenJDK 64-Bit Server VM (build 22, mixed mode, sharing)
```

Below a screen shot from the VSCode tests window:

![Copie d'écran_20240611_191514](https://github.com/CycloneDX/specification/assets/102670102/c8f2a7b9-afd4-43e0-b119-ae4c0060d7e9)

jkowalleck commented:
The modernizations in the tests make this a mix of topics.
@Nicolas-Peiffer, could you split the PR in multiple parts? One is the actual feature, and then additional pull requests for the modernizations?

Nicolas-Peiffer commented:
On the principle, I agree that should be 2 PR : 

* one for the XML catalog, 
* one for the update of Java unit tests. 

I have to figure the best way to do that. What I suggest :

* I close and cancel current PR: https://github.com/CycloneDX/specification/pull/477
* I create a branch `pr-xml-catalog` dedicated for the XML catalog. I do a new PR with this branch. This PR should be merged after the first, but not mandatory.
* I create a branch `pr-update-java-utests` dedicated for the update of Java-cycloneDX. I do a new PR with this branch. This PR should be merged second, but not mandatory.

I close this PR to open new ones.
